pkg/crawler/crawler.go

```diff
@@ -262,3 +313,260 @@ func (c *Crawler) fetchSHA1(url string) ([]byte, error) {
+			filesLicenseMap[data.FileName] = data.License
+			files = append(files, data.FileName)
+		}
+	}
+
+	if len(filesLicenseMap) == 0 {
+		return nil
+	}
+
+	// classify licenses
+
+	// 1 minute is the timeout for license classification of a file
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
+	defer cancel()
+
+	// c.opt.Limit is the number of concurrent tasks spawned to process license files
+	errs := c.classifier.ClassifyLicensesWithContext(ctx, int(c.opt.Limit), files, true)
+	if len(errs) > 0 {
+		log.Println("errors in license classification ", errs)
+	}
+
+	// extract results
+	results := c.classifier.GetResults()
+	sort.Sort(results)
+
+	// process results to update the normalizedLicenseMap
+	if results.Len() > 0 {
+		for _, r := range results {
+			if licenseVal, ok := filesLicenseMap[r.Filename]; ok {
+				// since results are sorted, we can skip processing of data with confidence <90%
+				if r.Confidence < 0.9 {
+					break
+				}
+
+				// skip processing since a higher confidence result is already processed
+				if licenseVal.ClassificationConfidence > r.Confidence {
+					continue
+				}
+
+				licenseVal.ClassificationConfidence = r.Confidence
+				filesLicenseMap[r.Filename] = licenseVal
+
+				// update normalized license map
+				normalizedLicenseMap[licenseVal.LicenseKey] = r.Name
+			}
+		}
+	}
+
+	defer func() {
+		// update normalized license map for license keys which couldn't be classified or had no url in pom for classification
+		uniqLicenseKeys := c.uniqueLicenseKeys.Items()
+		for key, license := range uniqLicenseKeys {
+			if _, ok := normalizedLicenseMap[key]; !ok {
+				if len(license.Name) > 0 {
+					normalizedLicenseMap[key] = license.Name
+				}
+			}
+		}
+
+		err := fileutil.WriteJSON(c.licensedir+types.NormalizedlicenseFileName, normalizedLicenseMap)
+		if err != nil {
+			log.Println(err)
+		}
+	}()
+
+	return nil
+}
+
+func (c *Crawler) prepareClassifierData(ctx context.Context) ([]licenseFilesMeta, error) {
+	log.Println("Preparing license classifier data")
+
+	var licenseFiles []licenseFilesMeta
+
+	// switch from concurrent to normal map
+	uniqLicenseKeyMap := c.uniqueLicenseKeys.Items()
+	uniqueLicenseKeyList := c.uniqueLicenseKeys.Keys()
+
+	client := http.Client{
+		Timeout: 10 * time.Second,
+	}
+
+	licenseKeyChannel := make(chan string, len(uniqueLicenseKeyList))
+
+	log.Printf("Total license keys to be processed %d", len(uniqueLicenseKeyList))
+
+	// dump license keys to the channel so that they can be processed
+	for _, key := range uniqueLicenseKeyList {
+		licenseKeyChannel <- key
+	}
+
+	limit := semaphore.NewWeighted(c.opt.Limit)
+
+	// error channel
+	errCh := make(chan error)
+	defer close(errCh)
+
+	// status channel to track processing of license keys
+	type status struct {
+		Meta licenseFilesMeta
+		Done bool
+	}
+	prepStatus := make(chan status, len(uniqueLicenseKeyList))
+	defer close(prepStatus)
+
+	// process license keys channel
+	go func() {
```

Comment:
I don't think we need concurrency for classifying licenses. We can fetch licenses when crawling URLs in parallel similar to sha1.

Reply:
This is a side effect of google license classifier usage. If we were to only process content in pom.xml licenses header directly, this won't be needed as you mentioned.

We are deferring the process of license classification so that we don't repeat it while crawling URLs and sha. Unique URLs/name across maven are around 2k. We'll be unncessarily repeating the classification process unless we introduce some thread safe cache which might slow down crawl due to locking.

We are trying to aggregate unique license urls from pom.xml, dump license url content in files concurrently and later on process them as a batch using google license classifier to get standardised license strings

Reply:
>unless we introduce some thread safe cache which might slow down crawl due to locking.

Using cache looks much simpler.

Reply:
Yup simpler, but slower. Crawling to concurrently process maven indexes will slow down due to blocking lock/unlock while reading/writing cache

Reply:
Did you measure it? I don't think it is so slow.